inc/class-job.php

```diff
@@ -186,39 +194,184 @@ public static function get_by_site( $site, $include_completed = false, $include_
+			'order' => 'ASC',
+			'__raw' => false,
+		];
+
+		$args = wp_parse_args( $args, $defaults );
+
+		// Allow passing a site object in
+		if ( is_object( $args['site'] ) && isset( $args['site']->blog_id ) ) {
+			$args['site'] = $args['site']->blog_id;
 		}
 
-		if ( empty( $results ) ) {
-			return [];
+		if ( ! is_numeric( $args['site'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_site_id' );
+		}
+
+		if ( ! empty( $args['hook'] ) && ! is_string( $args['hook'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_hook_name' );
+		}
+
+		if ( ! is_array( $args['args'] ) && ! is_null( $args['args'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_event_arguments' );
+		}
+
+		if ( ! is_numeric( $args['limit' ] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_limit' );
+		}
+
+		if ( $args['timestamp'] === 'future' ) {
+			$timestamp = time();
+			$timestamp_compare = '>';
+		} elseif ( $args['timestamp'] === 'past' ) {
+			$timestamp = time();
+			$timestamp_compare = '<=';
+		}
+
+		$args['limit' ] = absint( $args['limit' ] );
+
+		if ( $args['limit'] > 100 ) {
+			trigger_error( 'Exceeding recommended job search limit of 100' );
+		}
+
+		// Find all scheduled events for this site
+		$table = static::get_table();
+
+		$sql = "SELECT * FROM `{$table}` WHERE site = %d";
+		$sql_params[] = $args['site'];
+
+		if ( is_string( $args['hook'] ) ) {
+			$sql .= ' AND hook = %s';
+			$sql_params[] = $args['hook'];
+		}
+
+		if ( ! is_null( $args['args'] ) ) {
+			$sql .= ' AND args = %s';
+			$sql_params[] = serialize( $args['args'] );
+		}
+
+		if (
+			! empty( $args['timestamp'] ) &&
+			in_array( $args['timestamp'], ['past', 'future'] ) &&
+			! empty( $timestamp ) &&
+			! empty( $timestamp_compare ) &&
+			in_array( $timestamp_compare, [ '<=', '<', '>', '>=', '=' ] )
+		) {
+			$sql .= " AND nextrun {$timestamp_compare} %s";
+			$sql_params[] = date( DATE_FORMAT, (int) $timestamp );
+		} elseif (
+			! empty( $args['timestamp'] ) &&
+			is_array( $args['timestamp'] ) &&
+			count( $args['timestamp'] ) === 2
+		) {
+			// It's a range.
+			$sql .= ' AND nextrun BETWEEN %s AND %s';
+			$sql_params[] = date( DATE_FORMAT, (int) $args['timestamp'][0] );
+			$sql_params[] = date( DATE_FORMAT, (int) $args['timestamp'][1] );
+		} elseif ( ! empty( $args['timestamp'] ) ) {
+			$sql .= ' AND nextrun = %s';
+			$sql_params[] = date( DATE_FORMAT, (int) $args['timestamp'] );
+		}
+
+		$sql .= ' AND status IN(' . implode( ',', array_fill( 0, count( $args['statuses'] ), '%s' ) ) . ')';
+		$sql_params = array_merge( $sql_params, $args['statuses'] );
+
+		$sql .= ' ORDER BY nextrun';
+		if ( $args['order'] === 'DESC' ) {
+			$sql .= ' DESC';
+		} else {
+			$sql .= ' ASC';
+		}
+
+		if ( $args['limit'] > 0 ) {
+			$sql .= ' LIMIT %d';
+			$sql_params[] = $args['limit'];
+		}
+
+		// Cache results.
+		$last_changed = wp_cache_get_last_changed( 'cavalcade-jobs' );
+		$query_hash = md5( serialize( [ $sql, $sql_params ] ) ) . "::{$last_changed}";
```

Comment:
Let's avoid using `md5` in new code.

Reply:
Also, we need to double-check the LRU cache eviction on our cache servers currently, as I'm not sure if we're doing this.